src/IP.php

```diff
@@ -6,252 +6,402 @@
-	 * in PHP 5.2 (which I need to support).
-	 *
-	 * @return int
-	 */
-	abstract public function getVersion();
-
-	/**
-	 * Check if the IP is contained in given block.
-	 *
-	 * @param $block mixed Anything that can be converted into an IPBlock
-	 * @return bool
-	 */
-	public function isIn($block)
-	{
-		if ( ! $block instanceof IPBlock ) {
-			$block = IPBlock::create($block);
-		}
-
-		return $block->contains($this);
-	}
-
-	abstract public function isPrivate();
-
-	/**
-	 * Return true if the address is allocated for public networks
-	 *
-	 * @return bool
-	 */
-	public function isPublic()
-	{
-		return ! $this->isPrivate();
-	}
-}
\ No newline at end of file
+    const MAX_INT = null;
+    const NB_BITS = null;
+    const IP_VERSION = null;
+
+    /**
+     * Internal representation of the IP as a numeric format.
+     * For IPv4, this will be an SIGNED int (32 bits).
+     * For IPv6, this will be a GMP resource (128 bits big int).
+     *
+     * @var mixed
+     */
+    protected $ip;
+
+    /**
+     * @var bool
+     */
+    protected $is_private;
+
+    /**
+     * @var string Either "IPv4" or "IPv6"
+     */
+    protected $class;
+
+    /**
+     * Constructor tries to guess what is the $ip.
+     *
+     * @param $ip mixed String, binary string, int or float
+     */
+    public function __construct($ip)
+    {
+        if (is_int($ip)) {
+            $this->fromInt($ip);
+
+            return;
+        }
+
+        // float (or double) with an integer value
+        if (is_float($ip) && floor($ip) == $ip) {
+            $this->fromFloat($ip);
+
+            return;
+        }
+
+        if (is_string($ip)) {
+            $this->fromString($ip);
+
+            return;
+        }
+
+        if ((is_resource($ip) && 'GMP integer' === get_resource_type($ip)) || $ip instanceof \GMP) {
+            $this->fromGMP($ip);
+
+            return;
+        }
+
+        throw new \InvalidArgumentException(sprintf('Unsupported argument type: "%s".', gettype($ip)));
+    }
+
+    /**
+     * @param int $ip
+     */
+    private function fromInt($ip)
+    {
+        $ip = gmp_init(sprintf('%u', $ip), 10);
+
+        if (gmp_cmp($ip, static::MAX_INT) > 0) {
+            throw new \InvalidArgumentException(sprintf('The integer "%s" is not a valid IPv%d address.', gmp_strval($ip), static::IP_VERSION));
+        }
+
+        $this->ip = $ip;
+    }
+
+    /**
+     * @param float $ip
+     */
+    private function fromFloat($ip)
+    {
+        $ip = gmp_init(sprintf('%s', $ip), 10);
+
+        if (gmp_cmp($ip, 0) < 0 || gmp_cmp($ip, static::MAX_INT) > 0) {
+            throw new \InvalidArgumentException(sprintf('The double "%s" is not a valid IPv%d address.', gmp_strval($ip), static::IP_VERSION));
+        }
+
+        $this->ip = $ip;
+    }
+
+    /**
+     * @param string $ip
+     */
+    private function fromString($ip)
+    {
+        // binary, packed string
+        if (false !== @inet_ntop($ip)) {
+            $strLen = static::NB_BITS/8;
```

Comment:
This doesn't need to be a variable (I don't see it used anywhere else), and also we can have `static::NB_BYTES` as a constant instead of doing the calculation all the time.